alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/AbstractNavigationAction.kt

```diff
@@ -0,0 +1,250 @@
+         */
+        MOVING_TO_CROSSING_POINT_2,
+        /**
+         * When the second crossing point [isReached] (see [crossDoor]), the pedestrian may still be outside
+         * any room. In such case it moves towards [expectedNewRoom] centroid until he/she enters a room.
+         */
+        CROSSING_DOOR,
+        /**
+         * Moving to the final destination, which is inside [currentRoom] (this means it can be directly
+         * approached as no obstacle is placed in between).
+         */
+        MOVING_TO_FINAL,
+        ARRIVED
+    }
+    protected var state: State = State.START
+    /**
+     * Caches the room the pedestrian is into when he/she starts moving.
+     * When the pedestrian is crossing a door, it contains the room being left. When in [State.MOVING_TO_FINAL],
+     * it contains the room the pedestrian was (and should be) into. It's used to detect if the pedestrian ended
+     * up in an unexpected room while moving.
+     */
+    protected var previousRoom: Optional<M> = Optional.empty()
+    /**
+     * Defined when crossing a door. See [crossDoor].
+     */
+    protected var crossingPoints: Optional<Pair<P, P>> = Optional.empty()
+    /**
+     * Defined when crossing a door.
+     */
+    protected var expectedNewRoom: Optional<M> = Optional.empty()
+    /**
+     * Defined in [State.MOVING_TO_FINAL].
+     */
+    protected var finalDestination: Optional<P> = Optional.empty()
+
+    /**
+     * @returns the value if present or throws an [IllegalStateException] with a meaningful message.
+     */
+    protected fun <T> Optional<T>.orFail(): T = orElseThrow {
+        IllegalStateException("internal error: variable must be defined in $state")
+    }
+
+    /**
+     * Updates [pedestrianPosition] and [currentRoom], this can be costly.
+     * Depending on how [ConvexGeometricShape.contains] manage points on the boundary, the pedestrian could
+     * be inside two (adjacent) rooms at once. This can happen in two cases:
+     * - when in [State.MOVING_TO_CROSSING_POINT_1] or [State.MOVING_TO_FINAL] and the pedestrian is moving
+     * on [previousRoom]'s boundary. In such case [previousRoom] is used.
+     * - when crossing a door or in [State.NEW_ROOM] and [expectedNewRoom] is adjacent to [previousRoom].
+     * In such case [expectedNewRoom] is used.
+     * Otherwise the first room containing [pedestrianPosition] is used.
+     */
+    protected open fun updateCachedVariables() {
+        pedestrianPosition = environment.getPosition(pedestrian)
+        currentRoom = when {
+            (state == State.MOVING_TO_CROSSING_POINT_1 || state == State.MOVING_TO_FINAL) &&
+                previousRoom.orFail().contains(pedestrianPosition) -> previousRoom
+            (state == State.MOVING_TO_CROSSING_POINT_2 || state == State.CROSSING_DOOR || state == State.NEW_ROOM) &&
+                expectedNewRoom.isPresent &&
+                expectedNewRoom.get().contains(pedestrianPosition) -> expectedNewRoom
```

Comment:
It is not, but I though it was better to use optionals in general. Do they have any performance hit?

Reply:
Performance impact of optionals are irrelevant in this context.
Yes it's much better to use optionals over nullability in Java or Scala, since `null` eludes the type system (and by the way makes the entire type system unsound if used in conjunction with generics, see [this paper](https://dl.acm.org/doi/10.1145/2983990.2984004)). Kotlin takes a different stance (inspired by Ceylon) and provides nullable types. It's very much idiomatic to use them.

As a rule of thumb: if the consumer of your code may well be Java or Scala, prefer Optional. If your consumer is meant to be Kotlin (only or largely) then favor nullable types.

Reply:
Perfect, thanks for the explanation. I'll replace optionals with nullables.